### ISSUE TITLE
Make write transaction opening more safe

### ISSUE THREAD

**Issue by Kerollmops** [Kerollmops/heed#20]
Heed could ensure that only one write transaction is ever opened on the same thread.

It can create a [`thread_local`](https://doc.rust-lang.org/std/macro.thread_local.html) atomic counter for write transactions and raise an error (panic or not) when the user try to open a write transaction and another one is already open.

**Comment by ShadowJonathan** [comment]
How would this work with `Send` and the likes?

**Comment by Kerollmops** [comment]
Related to #148.

**Comment by Kerollmops** [comment]
Hey @ShadowJonathan,

> How would this work with Send and the likes?

Sorry for the delay. This is indeed something that I keep in mind. However, write transactions are non-`Send` (can't be moved between threads) and non-`Sync` (can't be referenced from another thread).

I also have some tricky questions for @hyc (👋) about that:
 - Is it allowed to have multiple read/write transactions opened on the same thread if they doesn't refer to the same environment?
 - If `MDB_NOTLS` is enabled, is it allowed to have one write transaction on a thread and multiple read transactions on this same thread? [The documentation explains](http://www.lmdb.tech/doc/group__mdb__env.html#ga5dca84a576d14b4bfe2deddc2dc622d3) that the read transaction will be stored in the `MDB_txn` struct and not a thread-local variable. Therefore I would answer _yes_ to my question.
 - Can you also confirm that when `MDB_NOTLS` is enabled, we can _move_ the read transaction between threads and also refer to them from other threads?

**Comment by hyc** [comment]
1) yes. different environments know nothing about each other and don't interact with each other.
2) yes but it is nonsensical to have multiple read txns on a single thread.
3) yes but a txn can only be used by one thread at a time, regardless. concurrent use of a single txn from multiple threads will cause memory corruption.

A thread a single unit of execution. So is a transaction. It is nonsensical to associate them other than 1:1.

**Comment by Kerollmops** [comment]
> FWIW that association can easily be broken when async comes in the mix, Send or not, because then a thread can consequently open multiple transactions on the same thread, and work with them on their own dime.

Indeed, this is why I would like to add a runtime check to only allow one transaction on a thread, write transactions are neither `Send` nor `Sync`. For read transactions, I would like to either always enable `MDB_NOTLS`, but I don't know if it can bring some limitation with other compile time options or keep track of the number of opened transactions too. Note that if the `sync-read-txn` heed feature is not enabled then the `RoTxn` is neither `Send` nor `Sync` and then it is safe as it is impossible to do a `.await` when a transaction is live.
